File: problems/C++/0033_sort_linked_list.cpp
//Problem Statement

/* Sort a linked list in O(n log n) time using constant space complexity. */
#include<bits/stdc++.h>

using namespace std;

struct ListNode {
   int val;
   ListNode *next;
   ListNode(int x) : val(x), next(NULL) {}
};

class Solution{
public:
  ListNode* sortList(ListNode* root){
    if(!root || !root->next)
      return root;

    ListNode* slow = root, *fast =root->next;

    while(fast && fast->next){
      fast = fast->next;
      if(fast)
        fast = fast->next;
      slow = slow->next;
    }

    ListNode *mid = slow->next;
    slow->next = NULL;
    return merge(sortList(root),sortList(mid));
  }

  ListNode *merge(ListNode * l1,ListNode *l2){
    ListNode * dummy = new ListNode(-1), *it = dummy;

    while(l1 && l2){
      if(l1->val<l2->val){
        it->next = l1;
        l1 = l1->next;
      }else{
        it->next = l2;
        l2 = l2->next;
      }

      it = it->next;
    }

    if(l1){
      it->next = l1;
    }else{
      it->next = l2;
    }

    return dummy->next;
  }
};

void printListNode(ListNode *root){
  if(!root)
    return;
  ListNode *it = root;
  while (it) {
    cout<<it->val<<"->";
    it=it->next;
  }
  cout<<"NULL"<<endl;
}

ListNode * vecToLinkedList(vector<int> vec){
  ListNode *dummy = new ListNode(-1), *it=dummy;

  for(auto val:vec){
    it->next = new ListNode(val);
    it = it->next;
  }

  return dummy->next;
}

int main(){

  Solution sol;
  vector<int> vec = {1,4,6,10,2,3,5,6,8,13};
  ListNode * mList = vecToLinkedList(vec);
  printListNode(mList);
  ListNode * sorted = sol.sortList(mList);
  printListNode(sorted);
  return 0;
}

File: Kotlin/src/main/kotlin/org/algorithm/problems/0033_sort_linked_list.kt
//Problem Statement

/* Sort a linked list in O(n log n) time using constant space complexity. */

package org.algorithm.problems

import org.algorithm.shared.ListNode

class `0033_sort_linked_list` {
    fun sortList(head: ListNode?): ListNode? {

        if (head == null)
            return head;

        var helper: MutableList<ListNode?> = mutableListOf<ListNode?>();
        var iterator: ListNode? = head;

        while (iterator != null) {
            var prev: ListNode?;
            prev = iterator;
            iterator = iterator.next;
            prev.next = null;
            helper.add(prev);
        }

        while (helper.size > 1) {
            var l1: ListNode? = helper[0];
            helper.removeAt(0);
            var l2: ListNode? = helper[0];
            helper.removeAt(0);
            var result: ListNode? = merge(l1, l2);
            helper.add(result);
        }

        return helper[0];
    }

    fun merge(l1: ListNode?, l2: ListNode?): ListNode? {
        val dummy: ListNode? = ListNode(-1);
        var iterator: ListNode? = dummy;
        var itl1: ListNode? = l1;
        var itl2: ListNode? = l2;

        while (itl1 != null && itl2 != null) {
            if (itl1.`val` < itl2.`val`) {
                iterator?.next = itl1;
                itl1 = itl1.next;
            } else {
                iterator?.next = itl2;
                itl2 = itl2.next;
            }

            iterator = iterator?.next;
        }

        if (itl1 == null)
            iterator?.next = itl2;
        else
            iterator?.next = itl1;

        return dummy?.next;
    }
}